allennlp/models/reading_comprehension/document_qa.py

```diff
@@ -0,0 +1,401 @@
+            eye = eye.cuda()
+        self_mask = Variable(eye).resize(1, passage_length, passage_length)
+        mask = mask * (1 - self_mask)
+
+        self_atten_probs = util.last_dim_softmax(self_atten_matrix, mask)
+
+        # Batch matrix multiplication:
+        # (batch, passage_len, passage_len) * (batch, passage_len, dim) -> (batch, passage_len, dim)
+        self_atten_vecs = torch.matmul(self_atten_probs, residual_layer)
+
+        # (extended_batch_size, passage_length, embedding_dim * 3)
+        concatenated = torch.cat([self_atten_vecs, residual_layer, residual_layer * self_atten_vecs],
+                                 dim=-1)
+
+        # _merge_self_atten => (extended_batch_size, passage_length, embedding_dim)
+        residual_layer = F.relu(self._merge_self_atten(concatenated))
+
+        # print("residual", residual_layer.size())
+
+        final_merged_passage += residual_layer
+        final_merged_passage = self._dropout(final_merged_passage)
+
+        # Bi-GRU in paper
+        start_rep = self._span_start_encoder(final_merged_passage, passage_lstm_mask)
+        span_start_logits = self._span_start_predictor(start_rep).squeeze(-1)
+        span_start_probs = util.masked_softmax(span_start_logits, passage_mask)
+
+        end_rep = self._span_end_encoder(torch.cat([final_merged_passage, start_rep], dim=-1), passage_lstm_mask)
+        span_end_logits = self._span_end_predictor(end_rep).squeeze(-1)
+        span_end_probs = util.masked_softmax(span_end_logits, passage_mask)
+
+        span_start_logits = util.replace_masked_values(span_start_logits, passage_mask, -1e7)
+        span_end_logits = util.replace_masked_values(span_end_logits, passage_mask, -1e7)
+
+        paragraph_span_start_logits = unsquash(span_start_logits, batch_size, num_paragraphs)
+        paragraph_span_end_logits = unsquash(span_end_logits, batch_size, num_paragraphs)
+
+        best_paragraph_word_span = self._get_best_span(paragraph_span_start_logits, paragraph_span_end_logits)
+
+        output_dict = {
+                "span_start_logits": span_start_logits,
+                "span_start_probs": span_start_probs,
+                "span_end_logits": span_end_logits,
+                "span_end_probs": span_end_probs,
+                "best_span": best_paragraph_word_span
+        }
+
+        if spans is not None:
+            # (batch_size, num_paragraphs, num_spans, 2)
+            span_idx_mask = 1 - torch.eq(spans, -1).float()
+            # (batch_size, num_paragraphs, num_spans)
+            span_idx_mask = span_idx_mask.max(dim=-1)[0]
+
+            # (batch_size, num_paragraphs, num_spans)
+            span_starts = spans[:, :, :, 0]
+            span_ends = spans[:, :, :, 1]
+
+            # (batch_size, num_paragraphs, num_tokens)
+            paragraphs_mask = util.get_text_field_mask(paragraphs, num_wrapping_dims=1).float()
+            # (batch_size, num_paragraphs)
+            paragraphs_mask = paragraphs_mask.max(dim=-1)[0]
```

Comment:
I don't understand this at all.
What is the scenario where this matrix is not all ones?
Does it check if a paragraph is completely empty, surely there is no need to check this?

Reply:
We batch together several instances, and some may have more paragraphs than others.  This means that we need to pad the number of paragraphs to a fixed length for all instances in the batch, so, yes, there will be some paragraphs that are completely empty.

Reply:
Yeah of course! I don't know how I missed that.